File: shared/src/commonMain/kotlin/com/vnidrop/app/AppGraph.kt
package com.vnidrop.app

import com.vnidrop.app.core.CoreGateway
import com.vnidrop.app.core.CoreRepository
import com.vnidrop.app.diagnostics.DiagnosticsCoordinator
import com.vnidrop.app.diagnostics.createDiagnosticsTransport
import com.vnidrop.app.feature.approvals.ApprovalCoordinator
import com.vnidrop.app.feature.send.AppFilePreviewRepository
import com.vnidrop.app.feature.send.createPlatformPreviewStore
import com.vnidrop.app.logging.AppLogger
import com.vnidrop.app.platform.AppVisibility
import com.vnidrop.app.preferences.AppPreferencesDefaults
import com.vnidrop.app.preferences.AppPreferencesRepository
import com.vnidrop.app.preferences.createAppPreferencesDataStore
import com.vnidrop.app.ui.feedback.UiMessageController
import com.vnidrop.app.ui.theme.ThemeMode
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.flow.drop
import kotlinx.coroutines.flow.filter
import kotlinx.coroutines.launch

class AppGraph(
	val dependencies: AppDependencies,
	private val applicationScope: CoroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate),
	val coreRepository: CoreGateway = CoreRepository(),
) {
	val visibility = AppVisibility()
	val messages = UiMessageController()
	val filePreviewRepository = AppFilePreviewRepository(
		createPlatformPreviewStore(dependencies.environment.defaultCoreDataDir),
	)
	val preferencesRepository = AppPreferencesRepository(
		dataStore = createAppPreferencesDataStore(dependencies.environment.defaultCoreDataDir),
		defaults = AppPreferencesDefaults(
			username = dependencies.environment.defaultUsername,
			receiveFolder = dependencies.fileSystemService.defaultReceiveFolder(),
			themeMode = ThemeMode.System,
			notificationsEnabled = false,
			diagnosticsEnabled = false,
		),
	)
	val diagnostics = DiagnosticsCoordinator.create(
		appDataDir = dependencies.environment.defaultCoreDataDir,
		appVersion = dependencies.environment.appVersion,
		platform = dependencies.environment.name,
		preferencesRepository = preferencesRepository,
		scope = applicationScope,
		transport = createDiagnosticsTransport(
			appVersion = dependencies.environment.appVersion,
			platform = dependencies.environment.name,
			installIdProvider = { preferencesRepository.ensureDiagnosticsInstallId() },
		),
	)
	val approvalCoordinator = ApprovalCoordinator(
		repository = coreRepository,
		preferencesRepository = preferencesRepository,
		notifications = dependencies.localNotificationService,
		visibility = visibility,
		messages = messages,
		scope = applicationScope,
	)

	init {
		AppLogger.initialize(dependencies.environment.defaultCoreDataDir)
		diagnostics.start()
		applicationScope.launch {
			visibility.isForeground
				.drop(1)
				.filter { isForeground -> !isForeground }
				.collect { diagnostics.telemetry.flush() }
		}
	}

	fun close() {
		coreRepository.shutdown()
		applicationScope.cancel()
	}
}

File: apple/VniDrop/App/AppGraph.swift
import Foundation
import Combine

/// Object graph wiring the repositories and coordinators together, ported from
/// `AppGraph.kt`. Owned by the app root for the process lifetime.
@MainActor
final class AppGraph: ObservableObject {
	let dependencies: AppDependencies
	let coreRepository: CoreRepository
	let visibility = AppVisibility()
	let messages = UiMessageController()
	let preferencesRepository: AppPreferencesRepository
	let filePreviewRepository: FilePreviewRepository
	let approvalCoordinator: ApprovalCoordinator
	let transferNotificationCoordinator: TransferNotificationCoordinator

	init(dependencies: AppDependencies, coreRepository: CoreRepository? = nil) {
		self.dependencies = dependencies
		let coreRepository = coreRepository ?? CoreRepository()
		self.coreRepository = coreRepository
		self.filePreviewRepository = FilePreviewRepository(appDataDir: dependencies.environment.defaultCoreDataDir)
		self.preferencesRepository = AppPreferencesRepository(
			fallback: AppPreferencesDefaults(
				username: dependencies.environment.defaultUsername,
				receiveFolder: dependencies.fileSystemService.defaultReceiveFolder(),
				themeMode: .system,
				diagnosticsEnabled: false
			)
		)
		self.approvalCoordinator = ApprovalCoordinator(
			repository: coreRepository,
			notifications: dependencies.notificationService,
			visibility: visibility,
			messages: messages
		)
		self.transferNotificationCoordinator = TransferNotificationCoordinator(
			repository: coreRepository,
			notifications: dependencies.notificationService,
			visibility: visibility,
			messages: messages
		)
		AppLogger.info("lifecycle", "graph created", ["platform": dependencies.environment.name])
	}

	func close() {
		coreRepository.shutdown()
	}
}
